In [1]:
[ord(character) for character in "€uro"]

[8364, 117, 114, 111]

In [7]:
for character in "€uro":
    decimal_code_pt = ord(character)
    #binary_code_pt = bin(decimal_code_pt)[2:]
    print(f"{character} {decimal_code_pt:10d} {decimal_code_pt:20b}")

€       8364       10000010101100
u        117              1110101
r        114              1110010
o        111              1101111


In [12]:
for character in "€uro":
    decimal_code_pt = ord(character)
    print(f"{character}  {decimal_code_pt:4d}  {decimal_code_pt:016b}")

€  8364  0010000010101100
u   117  0000000001110101
r   114  0000000001110010
o   111  0000000001101111


In [13]:
"€uro".encode("utf-8")

b'\xe2\x82\xacuro'

In [14]:
type("€uro".encode("utf-8"))

bytes

In UTF-8, the string `"€uro"` takes `3, 1, 1, 1` bytes to store for `"€", "u", "r", "o"`, resp. Indeed, this can be verified by the next cell:

In [15]:
for character in "€uro":
    print(character, len(character.encode("utf-8")))

€ 3
u 1
r 1
o 1


Note that `"€"` cannot be encoded in `"ascii"`.
```python
for character in "€uro":
    print(character, len(character.encode("ascii")))
```
<br>

```
UnicodeEncodeError: 'ascii' codec can't encode character '\u20ac' in position 0: ordinal not in range(128)
```

## Left Shift `<<`
`a << n` is like `a * (2**n)` for all non-negative integers `n`.
- Don’t use the bit shift operators as a means of premature optimization in Python. You won’t see a difference in execution speed, but you’ll most definitely make your code less readable. This is because most compilers and interpreters today, including Python's, are quite capable of optimizing your code behind the scenes.

In [18]:
import random

**(?)** What about left shift for negative integers?

In [59]:
[-0b1101 << n for n in range(10)]

[-13, -26, -52, -104, -208, -416, -832, -1664, -3328, -6656]

**Note**.
Expressions like `0b1101` are not two-complement. They can only express non-negative integers; at most we can add a minus sign in the front to negate.

In [1]:
0b11111111

255

In [2]:
-0b1111111

-127

In [7]:
-0b10000000

-128

## Right Shift `>>`
`a >> n` is like `a // (2**n)` for all non-negative integers `n`.

In [57]:
random.sample(range(2**8), 2**8)

[55,
 200,
 56,
 242,
 219,
 201,
 63,
 1,
 156,
 231,
 137,
 62,
 25,
 67,
 94,
 41,
 249,
 125,
 136,
 235,
 196,
 203,
 3,
 53,
 79,
 98,
 255,
 204,
 50,
 188,
 225,
 46,
 131,
 113,
 252,
 162,
 39,
 205,
 93,
 106,
 81,
 103,
 141,
 100,
 105,
 194,
 237,
 8,
 191,
 168,
 85,
 213,
 10,
 208,
 12,
 183,
 92,
 142,
 147,
 206,
 89,
 179,
 192,
 229,
 139,
 29,
 83,
 120,
 95,
 13,
 23,
 80,
 69,
 253,
 175,
 207,
 176,
 218,
 116,
 138,
 27,
 143,
 220,
 250,
 123,
 177,
 158,
 70,
 169,
 181,
 72,
 135,
 133,
 190,
 88,
 173,
 75,
 254,
 109,
 104,
 185,
 152,
 154,
 87,
 64,
 28,
 170,
 155,
 9,
 0,
 247,
 217,
 216,
 36,
 101,
 212,
 157,
 245,
 199,
 74,
 184,
 178,
 60,
 102,
 182,
 127,
 130,
 108,
 110,
 224,
 77,
 246,
 222,
 45,
 124,
 76,
 5,
 21,
 84,
 160,
 35,
 198,
 99,
 132,
 16,
 149,
 112,
 71,
 186,
 174,
 54,
 148,
 145,
 48,
 211,
 34,
 65,
 251,
 78,
 82,
 107,
 221,
 96,
 227,
 172,
 230,
 49,
 20,
 236,
 66,
 129,
 51,
 166,
 42,
 68,
 7,
 47,
 26,
 243,
 11

In [43]:
random.seed(42)
for a in random.sample(range(2**8), 2**8):
    for n in random.sample(range(2**8), 2**8):
        left_shift = a >> n
        divide_by_power = a // (2**n)
        if left_shift != divide_by_power:
            print(f"{a} >> {n} Not equal to {a} // (2**{n})")

In [44]:
type((i for i in range(10)))

generator

In [31]:
for n in range(17):
    print(f"2 >> {n:2d} equals {2 >> n}")

2 >>  0 equals 2
2 >>  1 equals 1
2 >>  2 equals 0
2 >>  3 equals 0
2 >>  4 equals 0
2 >>  5 equals 0
2 >>  6 equals 0
2 >>  7 equals 0
2 >>  8 equals 0
2 >>  9 equals 0
2 >> 10 equals 0
2 >> 11 equals 0
2 >> 12 equals 0
2 >> 13 equals 0
2 >> 14 equals 0
2 >> 15 equals 0
2 >> 16 equals 0


In [36]:
for n in range(33):
    print(f"-2 >> {n:2d} equals {-2 >> n}")

-2 >>  0 equals -2
-2 >>  1 equals -1
-2 >>  2 equals -1
-2 >>  3 equals -1
-2 >>  4 equals -1
-2 >>  5 equals -1
-2 >>  6 equals -1
-2 >>  7 equals -1
-2 >>  8 equals -1
-2 >>  9 equals -1
-2 >> 10 equals -1
-2 >> 11 equals -1
-2 >> 12 equals -1
-2 >> 13 equals -1
-2 >> 14 equals -1
-2 >> 15 equals -1
-2 >> 16 equals -1
-2 >> 17 equals -1
-2 >> 18 equals -1
-2 >> 19 equals -1
-2 >> 20 equals -1
-2 >> 21 equals -1
-2 >> 22 equals -1
-2 >> 23 equals -1
-2 >> 24 equals -1
-2 >> 25 equals -1
-2 >> 26 equals -1
-2 >> 27 equals -1
-2 >> 28 equals -1
-2 >> 29 equals -1
-2 >> 30 equals -1
-2 >> 31 equals -1
-2 >> 32 equals -1


**(?)** Why are the previous cell's results always `-1`?

### Arithmetic and Logical Shift
**Definition**.
- **logical shift**, aka _unsigned right shift_ or _zero-fill right shift_, moves the entire binary sequence, including the sign bit to the right and fills the resulting vacancies on the leftmost positions with zeros ![](figs/rshift_logical.5ee25943b1a4.gif)<br>
  - Since the leftmost bit is always replaced by a zero in a logical shift, the result is always **non-negative** <br><br>
- **arithmetic shift**, aka _signed right shift_, shifts all bits to the right, the vacancies being replaced by zeros except at the leading bit which maintains the sign bit's value. ![](figs/rshift_arithmetic.990b7e40923a.gif)

| decimal | binary |
| ------- | ------ |
| `-100` | `10011100` |
| `28`  |  `00011100` |

**(?)** Quick quiz: Whose bitwise NOT equals `-100`?


<br>
<br>

01. only arithmetic shift
  - Python
02. both logical and arithmetic shift
  - Java
  - Javascript
  - Julia
  
Because I hardly use Java or Javascript, let's use Julia to compare with Python.

In [15]:
-100 >> 1

-50

In [16]:
!julia -e "println(-100 >>> 1)"

9223372036854775758


In [20]:
!julia -e "println(typeof(-100))"

Int64


In [19]:
!julia -e "println(Int32(-100) >>> 1)"

2147483598


Let's
- explore a little bit
- try to do the logical shift manually via string operation in Julia by ourselves

In [22]:
!julia -e "println(typemax(Int32), \"\n\", typemin(Int32))"

2147483647
-2147483648


In [23]:
!julia -e "println(string(Int32(-100), base=2))"

-1100100


In [24]:
!julia -e "println(bitstring(Int32(-100)))"

11111111111111111111111110011100


In [25]:
!julia -e "println(length(bitstring(Int32(-100))))"

32


In [30]:
with open("logical_shift_manully.jl", "w") as f:
    f.write("""
a = Int32(-100)
a_bstr = bitstring(a)
shifted_bstr = "0" * a_bstr[1:end-1]
shifted = parse(Int32, shifted_bstr; base=2)
println("shifted           = $shifted")
println("Int32(-100) >>> 1 = $(Int32(-100) >>> 1)")
    """)

In [31]:
!cat logical_shift_manully.jl


a = Int32(-100)
a_bstr = bitstring(a)
shifted_bstr = "0" * a_bstr[1:end-1]
shifted = parse(Int32, shifted_bstr; base=2)
println("shifted           = $shifted")
println("Int32(-100) >>> 1 = $(Int32(-100) >>> 1)")
    

In [32]:
!julia logical_shift_manully.jl

shifted           = 2147483598
Int32(-100) >>> 1 = 2147483598


Great. We have in some sense verified logical right shift.

In [ ]:
from ctypes import c_uint32

In [ ]:
c_uint32(-100)

In [ ]:
c_uint32(-100).value

In [18]:
# Oops: Forgot to use the attribute .value
c_uint32(-100).value >> 1

2147483598

In [13]:
2**32 >> 7

33554432

In [12]:
!julia -e "println(2^32 >>> 7)"

33554432


In [14]:
!julia -e "println(2^32 >> 7)"

33554432


## Bitwise NOT
### Why `~156` Equals `-157` But Not `99` ?

| decimal | binary |
| ------- | ------ |
| `156` | `0b10011100` |
| `99`  | `0b01100011` |

This might be able to be explained as follows:

In [8]:
~156

-157

In [1]:
~-100

99

In [7]:
len(bin(~-100)), len(bin(-100))

(9, 10)

In [6]:
print(f"{bin(~-100):>10}")
print(bin(-100))

 0b1100011
-0b1100100


In [17]:
0b01100011

99

In [14]:
ls_result = !ls
ls_result

['bitwise_ops.ipynb', 'figs', 'README.md', 'trash.py']

In [13]:
julia_result = !(julia -e "print(Int32(-100) >>> 1)")
julia_result

['2147483598']

### Unsigned Integer (in C)
- The first sign bit of a signed integer, when in unsigned case, is being recognized as an extra bit. In other words, the maximum reachable integer of an unsigned integer is larger.
- Python's integer can be infinite-length, so there is no worry about overflow.

In [1]:
from ctypes import c_uint8

In [3]:
c_uint8(-42).value

214

In [4]:
bin(42)

'0b101010'

So  `-42` by two's complement should be `~(0b00101010) + 0b1`, which equals `0b11010110`. When this is
interpreted as `c_uint8`, we get `2**7 + 2**6 + 2**4 + 2**2 + 2**1`, which equal to `128 + 64 + 16 + 4 + 2`,
i.e. `214`.

**(?)** What happens with overflow in `c_uint8`?<br>
**(R)** Cf. below.

In [5]:
c_uint8(0)

c_ubyte(0)

In [6]:
c_uint8(2**8)

c_ubyte(0)

In [7]:
c_uint8(2**8 + 1)

c_ubyte(1)

In [8]:
c_uint8(-1)

c_ubyte(255)

Expressing `-5` in Python in
- `4` bits
- `8` bits
- `5` bits(?)
- `7` bits(?)

In [14]:
f"{-5 & 0b1111:04b}"

'1011'

If we interpret `0b1011` using two's complement, then it is the negative of `0b0101` (because `0b1010 + 0b0001` equals `0b1011`), i.e. the negative of `5`. Therefore the result in Python is consistent with that of two's complement.

In [15]:
f"{-5 & 0b11111111:08b}"

'11111011'

Similarly, in two's complement, `011111011`, equal to `0b11111010 + 0b1`,  is the negative of `0b00000101`. So, in `8` bits, the two results are also consistent.

**(?)** Is it necessary to `& 0b1111`?<br>

In [10]:
f"{-5:04b}"

'-101'

In [11]:
f"{3:04b}"

'0011'

In [12]:
f"{3:03b}"

'011'

In [13]:
f"{-5:08b}"

'-0000101'

In [16]:
f"{-5:03b}"

'-101'

In [17]:
f"{-5:05b}"

'-0101'

**(R)** `& 0b1111` **is necessary**: As we can observe from the experimental cells above, using formated string with `{x:04b}` without `x & 0b1111`, when x is a negative integer, Python will **print the negative sign** and **count that as a character**.

In [18]:
f"{-5 & 0b1111:4b}"

'1011'

In [20]:
f"{-5 & 0b1111:5b}"

' 1011'

In [19]:
f"{-5 & 0b11111111:8b}"

'11111011'

In [21]:
f"{-5 & 0b11111111:7b}"

'11111011'

In [22]:
f"{-5 & 0b11111111:9b}"

' 11111011'

### One's Complement

To show that
> the binary sequences of negative numbers in one’s complement are arranged **in reverse order** as compared to sign-magnitude

We set out to prove that if `a` and `b` are two negative `8`-bit integers with `a > b`, then
> (`a` with the leftmost bit flipped to `0`) $<$ (`b` with the leftmost bit flipped to `0`)

**(?)** Is the carryover in one's complement means the following?
```
    0b11000001
+)  0b01000000
---------------
    0b00000010
```

### Two's Complement

### Floating-Point Numbers and Fixed-Point Arithmetics
The rounding errors of floating-point numbers is unacceptable when doing **monetary calculations**.

In [23]:
0.1 + 0.2

0.30000000000000004

Python's `decimal` module allows one to do fixed-point arithmetics.

In [24]:
from decimal import Decimal, localcontext

In [25]:
with localcontext() as context:
    context.prec = 5  # Number of digits
    print(Decimal("123.456"))
    print(Decimal("123.456")*1)

123.456
123.46


If one does not like to use fixed-point arithmetics, one can also, for example, for US dollars, use the smallest unit `cents` and use **integer**, instead of using floating/fixed-point arithmetics along with `dollars`.

## Integer Representations in Python

HTML still has its advantage over Markdown:
- <code>128<sub>10</sub></code> vs `(128)_(10)` vs $(128)_{10}$
- `<code>128<sub>10</sub></code>` vs `` `(128)_(10)` `` vs `$(128)_{10}$`

### Interned Integers
> "In CPython, very small integers between <code>-5<sub>10</sub></code> and <code>256<sub>10</sub></code> are interned in a global cache to gain some performance because numbers in that range are commonly used. In practice, whenever you refer to one of those values, which are singletons created at the interpreter startup, Python will always provide the same instance." -- Bartosz Zaczyński (from Real Python)

In [26]:
a = 256
b = 256
a is b

True

In [27]:
print(id(a), id(b), sep="\n")

94339986811552
94339986811552


In [31]:
def non_negative_int_generator():
    n = 0
    while True:
        yield n
        n += 1

for n in non_negative_int_generator():
    a = n
    b = n
    if a is not b:
        print(f"n = {n}: a is not b")
        break
    if n > 1e6:
        break

**(?)** Note that the above cell does not allow us to find the upper bound for the interned integers. How can we modify it to achieve this goal?

In [37]:
a = -5
b = -5
a is b

True

In [40]:
a = 257
b = 257
a is b

False

In [41]:
print(id(a), id(b), sep="\n")

140357728138608
140357728138832


In [42]:
a = -6
b = -6
a is b

False

In [35]:
a = 1000
b = 1000
a is b

False

In [43]:
print(id(257), id(257), sep="\n")

140357728138832
140357728138832


In [45]:
print(id(257), id(257), sep="\n")

140357728139056
140357728139056


In [46]:
print(id(257), id(257), sep="\n")

140357728138480
140357728138480


Note how the ids are
- equal inside each cell
- unequal btw distinct cells

###  Fixed-Precision Integers
`sys.maxsize` will tell which architecture you computer is, normally btw

- `32`-bit architecture
- `64`-bit architecture

In [47]:
import sys
sys.maxsize

9223372036854775807

In [48]:
import math
math.log2(sys.maxsize)

63.0

As far as I can understand, this means that in Python integers **from `-sys.maxsize` to `sys.maxsize - 1` (inclusive)** are stored in **`lg(sys.maxsize)` bit** using **two's complement**. For example, on `64`-bit architecture, this means that integers $\in [-2^{63}, 2^{63} - 1]$ are stored as a sequence of `64` zeros and ones using two's complement.

What about integers beyond that range? Well, it seems that Python stores them in a different way. Cf. ![next section](#arbitrary-precision-integers)

### Arbitrary-Precision Integers

In [51]:
math.factorial(42)

1405006117752879898543142606244511569936384000000000

In [50]:
_.bit_length()

170

Note that the `.bit_length()` method of the `int` class gives the same result as
```python
len(bin(_)) - 2
```

This means it takes `170` bits to store this number, way beyond `64` bits.

Here is my understanding of this part:

01. Big numbers are stored in Python3 using **sign magnitude** and with base <b><code>2<sup>30</sup></code></b>
02. Performing bignum arithmetics is slow
03. Conversion is needed when doing arithmetics btw a bignum and a fixed-precision integer because one is under sign magnitude, the other under two's complement

In [54]:
sys.int_info

sys.int_info(bits_per_digit=30, sizeof_digit=4)

**(?)** Explain the meaning of `bits_per_digit` and `sizeof_digit`.<br>

## Manipulating Numbers with Diff Bases in Python

> "_When you call_ `bin()` _on a negative integer, it merely prepends the minus sign to the bit string obtained from the corresponding positive value_"

As a consequence, `bin(n)` where `n` is a negative integer, the returned string is more like a sign-magnitude result.

In [55]:
print(bin(-42), bin(42), sep="\n ")

-0b101010
 0b101010


In [56]:
print(int("-101010", 2))   # -42, like sign-magnitude-like
print(int("00101010", 2))  # 42
print(int("11010110", 2))  # ?

-42
42
214


In [57]:
print(int("0b00101010", 2))  # 42
print(int("0b11010110", 2))  # Two's complement?

42
214


In [58]:
print(0b00101010)
print(0b11010110)

42
214


In [59]:
print(0b11010110 & 0b11111111)

214


**(?)** Why not `-42`?

In [62]:
int("0b1" + "0"*63, 2)

9223372036854775808

In [63]:
_ == 2**63

True

In [65]:
0b1000_0000000000_0000000000_0000000000_0000000000_0000000000_0000000000

9223372036854775808

In [66]:
_ == 2**63

True

Regardless of which implementation (two's complement, sign magnitude, one's complement, etc.) internally Python might use (to represent its integers), the interface Python provides us (e.g. `42`, `0b101010`, `0x2a`, etc.) seems to resemble more **sign magnitude**. Indeed, as we can observe

- `int("0b1" + "0"*63, 2)` is not negative
- The two's complement of `0b101010` (which is <code>42<sub>10</sub></code>) should have been `0b11010110`; however in Python `0b11010110` gives `214`.

In other words, expressions like integer literals, `0b101010111`, `0xaecd091`, etc. can only express **positive integers**. The **negative sign** must be specified via the symbol **`-`**.

### Emulate Fixed-Length Bit Sequences Containing the Sign Bit
Via, e.g. one of the following

- Bitmask
- Modulo operation (`%`)
- `ctypes` module
- `array` module
- `struct` module


In [68]:
mask = 0xff
bin(-42 & mask)

'0b11010110'

In [69]:
-42 & mask

214

What happened?
> _Masking forces Python to temporarily change the number’s representation<br>
> from sign-magnitude to two’s complement and then back again._

In [70]:
bin(-42)

'-0b101010'

In [71]:
bin(-42 % (1 << 8))  # Give me eight bits

'0b11010110'

Maybe thinking in this way will help you clarify:
> - Internally, Python uses two's complement for signed integers of fixed length
> - But, don't think of expressions like literal integer (e.g. `42`), binary expressions (e.g. `0b101010`), etc. as two's complement; they are more like sign magnitude
> - When we do bitwise operations, it **still obeys the two's complement** rules, but the final result will be understood as sign magnitude, i.e. as being **w/o sign bit**, or in other words, **positive**.